types/react/index.d.ts

```diff
@@ -558,7 +580,21 @@ declare namespace React {
 
     function createRef<T>(): RefObject<T>;
 
-    function forwardRef<T, P = {}>(Component: RefForwardingComponent<T, P>): ComponentType<P & ClassAttributes<T>>;
+    function forwardRef<T, P = {}>(Component: RefForwardingComponent<T, P>): ExoticComponent<P & ClassAttributes<T>>;
+
+    type ComponentProps<T extends ComponentType<any>> =
+        T extends ComponentType<infer P> ? P : {};
+    type ComponentPropsWithRef<T extends ComponentType<any>> =
+        T extends ComponentClass<infer P>
+            ? P & ClassAttributes<InstanceType<T>>
+            : T extends SFC<infer P>
+                ? P
+                : {};
+
+    function memo<T extends ComponentType<any>>(
```

Comment:
Both `memo` and `forwardRef` will break the `LibraryManagedAttributes` inference, since the static properties are not copied over. Is there any way we can preserve those when using these "HOC-like" functions?

Reply:
I tried to in my follow-up PR, but it seems `LibraryManagedAttributes` just don't work at all with non-class components, even if the type that `JSX.LibraryManagedAttribute` extracts is correct.

Reply:
Is this still the case? AFAIK it should apply correctly to all components now.